internal/driver/config/provider.go

```diff
@@ -131,66 +134,68 @@ func (k *Provider) CORS() (cors.Options, bool) {
 	})
 }
 
-func (k *Provider) DSN() string {
+func (k *Config) DSN() string {
 	dsn := k.p.String(KeyDSN)
 	if dsn == "memory" {
 		return DSNMemory
 	}
 	return dsn
 }
 
-func (k *Provider) TracingServiceName() string {
+func (k *Config) TracingServiceName() string {
 	return k.p.StringF("tracing.service_name", "ORY Keto")
 }
 
-func (k *Provider) TracingProvider() string {
+func (k *Config) TracingProvider() string {
 	return k.p.StringF("tracing.provider", "")
 }
 
-func (k *Provider) TracingConfig() *tracing.Config {
+func (k *Config) TracingConfig() *tracing.Config {
 	return k.p.TracingConfig("ORY Keto")
 }
 
-func (k *Provider) NamespaceManager() (namespace.Manager, error) {
-	if k.nm == nil {
-		k.nmLock.Lock()
-		defer k.nmLock.Unlock()
-
-		var ctx context.Context
-		ctx, k.cancelNamespaceManager = context.WithCancel(k.ctx)
-
-		switch nTyped := k.p.GetF(KeyNamespaces, "file://./keto_namespaces").(type) {
-		case string:
-			var err error
-			k.nm, err = NewNamespaceWatcher(ctx, k.l, nTyped)
-			if err != nil {
-				return nil, err
-			}
-		case []*namespace.Namespace:
-			k.nm = NewMemoryNamespaceManager(nTyped...)
-		case []interface{}:
-			nEnc, err := json.Marshal(nTyped)
-			if err != nil {
-				return nil, errors.WithStack(err)
-			}
-
-			nn := make([]*namespace.Namespace, len(nTyped))
-
-			if err := json.Unmarshal(nEnc, &nn); err != nil {
-				return nil, errors.WithStack(err)
-			}
-
-			k.nm = NewMemoryNamespaceManager(nn...)
-		default:
-			return nil, errors.Errorf("could not create namespace manager from %#v, this indicates an error in the JSON schema that should be reported", nTyped)
+func (k *Config) NamespaceManager() (namespace.Manager, error) {
+	if nm := func() namespace.Manager {
+		k.nmLock.RLock()
+		defer k.nmLock.RUnlock()
+		return k.nm
+	}(); nm != nil {
+		return nm, nil
+	}
+
+	fmt.Println("creating new nm")
+
+	k.nmLock.Lock()
+	defer k.nmLock.Unlock()
+
+	var ctx context.Context
+	ctx, k.cancelNamespaceManager = context.WithCancel(k.ctx)
```

Comment:
Do you have a test how this behaves at scale or with constant reloading? We should avoid any context or memory leaks.

Reply:
this code was just moved from https://github.com/ory/keto/pull/638/files#diff-27444dce264cbc384993274b30b80fcf763ad3255d0a7df124d20d2e047dad21L159-L160

Reply:
Question still remains :)

Reply:
No there is no specific test for this context. Tracked as #642